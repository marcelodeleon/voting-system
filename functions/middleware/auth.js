const jwt = require('jsonwebtoken');
const { User } = require('../../libs/models');
const { mongodb } = require('../../libs/connectors');

const jwtSecret = process.env.JWT_SECRET;
const mongodbUri = process.env.MONGODB_URI;

const verifyToken = (token) => {
  try {
    return jwt.verify(token, jwtSecret);
  } catch (err) {
    if (err.name === 'JsonWebTokenError') {
      return null;
    }

    throw err;
  }
};

const withAuth = (func) => (a, b) => {
  const handler = {
    apply: async (target, thisArg, args) => {
      await mongodb(mongodbUri);
      const event = args[0];
      const context = args[1];

      const { headers } = event;
      const { authorization } = headers;

      if (!authorization) {
        return {
          statusCode: 401,
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ error: 'You must be logged in.' }),
        };
      }

      const [scheme, token] = authorization.split(' ');

      if (!/^Bearer$/i.test(scheme)) {
        return {
          statusCode: 401,
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            error: `Unsupported authentication scheme: "${scheme}". Supported schemes: "Bearer".`,
          }),
        };
      }

      const decoded = verifyToken(token);
      if (!decoded) {
        return {
          statusCode: 401,
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            error: 'Invalid token.',
          }),
        };
      }

      const { sub: userId } = decoded;
      const user = await User.findById(userId);

      if (!user) {
        return {
          statusCode: 401,
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            error: 'Invalid token.',
          }),
        };
      }

      return target(event, context);
    },
  };

  const proxy = new Proxy(func, handler);

  return proxy.apply(this, [a, b]);
};

module.exports = withAuth;
