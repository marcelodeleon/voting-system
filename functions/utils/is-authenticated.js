const verifyToken = (token) => {
  try {
    return jwt.verify(token, jwtSecret);
  } catch (err) {
    if (err.name === 'JsonWebTokenError') {
      return null;
    }

    throw err;
  }
};

module.exports = (headers) => {};
