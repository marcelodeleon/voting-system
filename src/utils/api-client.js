const { NODE_ENV } = process.env;

async function request(resource, config) {
  const apiOrigin =
    NODE_ENV === 'development'
      ? 'http://localhost:8888/.netlify/functions/'
      : 'https://voting-system-tas.netlify.app/.netlify/functions/';

  const response = await fetch(`${apiOrigin}${resource}`, {
    ...config,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(config.body),
  });

  try {
    const body = await response.clone().json();
    console.log({ status: response.status });
    if (response.ok || (response.status >= 200 && response.status < 400))
      return body;
    console.log({ body });
    return Promise.reject(body.error);
  } catch (error) {
    if (response.status === 500)
      return Promise.reject(new Error('Internal Server Error'));
    return Promise.reject(error);
  }
}

const apiClient = {
  get: (url, config) => request(url, { ...config, method: 'GET' }),
  post: (url, config) => request(url, { ...config, method: 'POST' }),
  put: (url, config) => request(url, { ...config, method: 'PUT' }),
  patch: (url, config) => request(url, { ...config, method: 'PATCH' }),
  delete: (url, config) => request(url, { ...config, method: 'DELETE' }),
};

export default apiClient;
