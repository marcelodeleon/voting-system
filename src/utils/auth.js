import { getSessionToken } from './session';

// eslint-disable-next-line import/prefer-default-export
export const isAuthenticated = () => !!getSessionToken();
