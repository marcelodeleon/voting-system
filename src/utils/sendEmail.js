//require('dotenv').config();

const sgMail = require('@sendgrid/mail');

//sgMail.setApiKey(process.env.SENDGRID_API_KEY)

function unixTimestamp(date) {
  if (date === undefined) {
    date = new Date();
  }
  return Math.floor(date.getTime() / 1000.0);
}

sgMail.setApiKey(
  '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',
);

function sendEmail(emailTo, subject, html, sendAt) {
  msg = {
    to: emailTo,
    from: 'topicosavanzados2020.3@gmail.com',
    subject: subject,
    html: html,
    //the sendAt time is in UNIXTimestamp
    //https://www.epochconverter.com
    sendAt: unixTimestamp(sendAt),
  };
  sgMail
    .sendMultiple(msg)
    .then(() => {
      console.log(msg);
    })
    .catch((err) => {
      console.log(err);
      console.log(err.response.body);
    });
  // .catch((error) => callback(error));
}

module.exports = sendEmail;

//sendEmail('gonzalogg.garcia@gmail.com',new Date('2020-11-05T22:45:57.413Z'))
