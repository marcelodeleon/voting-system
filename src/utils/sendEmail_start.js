require('dotenv').config();
const sgMail = require('@sendgrid/mail');
const { NODE_ENV } = process.env;
//sgMail.setApiKey(process.env.SENDGRID_API_KEY)

const { urlOrigin } =
  NODE_ENV === 'development'
    ? 'http://localhost:8888/vote'
    : 'https://voting-system-tas.netlify.app/vote';

unixTimestamp: function unixTimestamp(date) {
  if (date === undefined) {
    date = new Date();
  }
  return Math.floor(date.getTime() / 1000.0);
}
sgMail.setApiKey(
  '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',
);
const msg = {
  to: 'gonzalogg.garcia@gmail.com',
  from: 'topicosavanzados2020@gmail.com',
  subject: 'Comienza el periodo de votacion',
  //html: html,
  html:
    '<strong>Comienza el período de votación, ingresa </strong><a href=http://localhost:8888/vote>aquí</a>',
  //the sendAt time is in UNIXTimestamp

  //https://www.epochconverter.com
  sendAt: 1604553497,
  //sendAt: unixTimeStamp(date),
};

sgMail
  .send(msg)
  .then(() => {
    console.log(msg);
  })
  .catch((error) => callback(error));
