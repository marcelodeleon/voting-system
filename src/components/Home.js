import React from 'react';
import { isAuthenticated } from '../utils/auth';
import LogIn from './users/LogIn';
import Welcome from './Welcome';

export default function Home() {
  return <div>{isAuthenticated() ? <Welcome /> : <LogIn />}</div>;
}
