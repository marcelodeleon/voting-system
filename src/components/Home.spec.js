import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import 'mutationobserver-shim';
import Home from './Home';
import { isAuthenticated } from '../utils/auth';

global.MutationObserver = window.MutationObserver;
jest.mock('../utils/auth');

describe('Home', () => {
  beforeEach(() => {
    isAuthenticated.mockRestore()
  })

  test('Renders Welcome screen if user authenticated correctly', () => {
    isAuthenticated.mockResolvedValue(true);
    const { container } = render(<Home />, { wrapper: MemoryRouter });
    expect(container.firstChild).toMatchSnapshot();
  });

  test.only('Renders LogIn screen if user is not authenticated', () => {
    const { container } = render(<Home />, { wrapper: MemoryRouter });
    expect(container.firstChild).toMatchSnapshot()
  });
});
