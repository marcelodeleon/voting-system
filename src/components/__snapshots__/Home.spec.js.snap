// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Renders LogIn screen if user is not authenticated 1`] = `
<div>
  <form
    class="sc-bdfBwQ eYhSFl"
  >
    <h1>
      Login
    </h1>
    <div
      class="sc-gsTCUz kXIVEF"
    >
      <label>
        Email:
        <input
          name="email"
          style="width: 30rem;"
          type="email"
          value=""
        />
      </label>
      <label>
        Password:
        <input
          name="password"
          style="width: 30rem;"
          type="password"
          value=""
        />
      </label>
      <input
        class="sc-dlfnbm WzKeF"
        type="submit"
        value="Log In"
      />
      <a
        href="/users/registration"
      >
        ¿No tienes cuenta? Click aquí para registrar un usuario
      </a>
    </div>
  </form>
</div>
`;

exports[`Home Renders Welcome screen if user authenticated correctly 1`] = `
<div>
  <div
    class="sc-iBPRYJ kEDcat"
  >
    <div
      class="sc-hKgILt bwBfkL Navbar"
    >
      <a
        class="Navbar__brand"
        href="/vote"
      >
        <img
          alt="logo"
          class="Navbar__brand-logo"
          src="logo.svg"
        />
      </a>
      <div
        class="sc-jSgupP ePKIeF"
      >
        <a
          class="sc-eCssSg cACQva"
          href="/"
        >
          Cerrar Sesión
        </a>
      </div>
    </div>
    <h1>
      Bienvenido al Sistema de Votación!
    </h1>
    <p>
      Recibirá un email con las votaciones a las que pueda participar en cuanto estén listas.
    </p>
    <p>
      Los links de votación son de un único uso y deberá haber iniciado sesión en la aplicación.
    </p>
  </div>
</div>
`;
